src/modules/frost/main_impl.h

```diff
@@ -0,0 +1,804 @@
+    secp256k1_frost_verify_share_ecmult_data *ctx = (secp256k1_frost_verify_share_ecmult_data *) data;
+    int ret;
+
+    ret = secp256k1_pubkey_load(ctx->ctx, pt, *(ctx->pubcoeff)+idx);
+    VERIFY_CHECK(ret);
+    secp256k1_scalar_mul(sc, &secp256k1_scalar_one, &ctx->idxn);
+    secp256k1_scalar_mul(&ctx->idxn, &ctx->idxn, &ctx->idx);
+
+    return 1;
+}
+
+static int secp256k1_frost_pubkey_combine_callback(secp256k1_scalar *sc, secp256k1_ge *pt, size_t idx, void *data) {
+    secp256k1_frost_pubkey_combine_ecmult_data *ctx = (secp256k1_frost_pubkey_combine_ecmult_data *) data;
+
+    secp256k1_scalar_set_int(sc, 1);
+    /* the public key is the first index of each set of coefficients */
+    return secp256k1_pubkey_load(ctx->ctx, pt, &ctx->pks[idx][0]);
+}
+
+int secp256k1_frost_share_agg(const secp256k1_context* ctx, secp256k1_frost_share *agg_share, secp256k1_xonly_pubkey *agg_pk, unsigned char *vss_hash, const secp256k1_frost_share * const* shares, const secp256k1_pubkey * const* pubcoeffs, uint16_t n_shares, uint16_t threshold, uint16_t my_index) {
+    secp256k1_frost_pubkey_combine_ecmult_data pubkey_combine_ecmult_data;
+    secp256k1_gej pkj;
+    secp256k1_ge pkp;
+    int pk_parity;
+    secp256k1_scalar acc;
+    uint16_t i;
+    int overflow;
+
+
+    VERIFY_CHECK(ctx != NULL);
+    ARG_CHECK(agg_share != NULL);
+    ARG_CHECK(agg_pk != NULL);
+    ARG_CHECK(vss_hash != NULL);
+    ARG_CHECK(shares != NULL);
+    ARG_CHECK(pubcoeffs != NULL);
+    ARG_CHECK(n_shares > 0);
+    ARG_CHECK(my_index > 0);
+
+    if (threshold == 0 || threshold > n_shares) {
+        return 0;
+    }
+
+    secp256k1_scalar_clear(&acc);
+    for (i = 0; i < n_shares; i++) {
+        secp256k1_scalar share_i;
+        secp256k1_frost_verify_share_ecmult_data verify_share_ecmult_data;
+        secp256k1_gej sharej;
+        secp256k1_gej expectedj;
+
+        secp256k1_scalar_set_b32(&share_i, shares[i]->data, &overflow);
+        if (overflow) {
+            return 0;
+        }
+
+        verify_share_ecmult_data.ctx = ctx;
+        verify_share_ecmult_data.pubcoeff = &pubcoeffs[i];
+        /* Evaluate the public polynomial at the index */
+        secp256k1_scalar_set_int(&verify_share_ecmult_data.idx, my_index);
+        secp256k1_scalar_set_int(&verify_share_ecmult_data.idxn, 1);
+        /* TODO: add scratch */
+        if (!secp256k1_ecmult_multi_var(&ctx->error_callback, NULL, &sharej, NULL, secp256k1_frost_verify_share_ecmult_callback, (void *) &verify_share_ecmult_data, threshold)) {
+            return 0;
+        }
+        /* Verify share using VSS */
+        secp256k1_ecmult_gen(&ctx->ecmult_gen_ctx, &expectedj, &share_i);
+        secp256k1_gej_neg(&expectedj, &expectedj);
+        secp256k1_gej_add_var(&expectedj, &expectedj, &sharej, NULL);
+        if (!secp256k1_gej_is_infinity(&expectedj)) {
+            return 0;
+        }
```

Comment:
Wouldn't you want to know which share does not pass the check in order to allow identifying disruptive signers?